geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java

```diff
@@ -245,33 +244,25 @@ static int setOpStoreResult(RegionProvider regionProvider, RedisKey destinationK
     return scanResult;
   }
 
-  public Collection<byte[]> spop(Region<RedisKey, RedisData> region, RedisKey key, int popCount) {
-    int originalSize = scard();
-    if (originalSize == 0) {
-      return emptyList();
-    }
-
-    if (popCount >= originalSize) {
-      region.remove(key, this);
-      return members;
-    }
-
-    List<byte[]> popped = new ArrayList<>();
-    byte[][] setMembers = members.toArray(new byte[originalSize][]);
-    Random rand = new Random();
-    while (popped.size() < popCount) {
-      int idx = rand.nextInt(originalSize);
-      byte[] memberToPop = setMembers[idx];
-      if (memberToPop != null) {
-        setMembers[idx] = null;
-        popped.add(memberToPop);
-        membersRemove(memberToPop);
+  public List<byte[]> spop(int count, Region<RedisKey, RedisData> region, RedisKey key) {
+    List<byte[]> result = new ArrayList<>();
+    if (members.size() <= count) {
+      result.addAll(members);
+      members.clear();
+    } else {
+      Random rand = new Random();
+      while (result.size() != count) {
+        int randIndex = rand.nextInt(members.getBackingArrayLength());
+        byte[] member = members.getFromBackingArray(randIndex);
+        if (member != null) {
```

Comment:
The load factor is 3/4, so I believe it would be null 1/4 of the time. I used this method to get set members from the backing array, because of Darrel's comment on not copying over the member set. 

I could be wrong, but I don't think it would be stuck in the loop infinitely. In SPOPEXECUTOR there is a check to make sure that the set size is greater than 0, and in spop there is a check before this that makes sure the count is less than the size of the set. So when entering this loop there would be some members in the set, and it would remove an amount that is less than the total size of the set. I guess it could loop infinitely if spop is called outside SPOPEXECUTOR, but I could move the 0 size check in this method.

Reply:
As you call "members.remove()" will that cause the backingArrayLength to eventually shrink?

Reply:
What if we changed our getFromBackingArray to not return null? If the initial index given finds a null it could just inc it until it finds a non-null or hits the end of the array. If it hits the end it could start at 0 and continue the linear scan. As long as we start at the random index given I think we will be honoring the randomness spop desires and since we remove the thing we find we will not keep returning the same member. Also maybe this should be "removeFromBackingArray(int)" instead of get. That way you don't need to turn around and call "remove(byte[])" which has to search for it.

Reply:
As I call members.remove() it will cause the backingArrayLength to shrink, and might possibly shrink multiple times if the count is large enough.

Reply:
The method it uses to remove an element at the specified index is private, so it would need to be reimplemented, but I am not sure if there is some legal issues with doing that.